client/trino-client/src/main/java/io/trino/client/JsonResponse.java

```diff
@@ -108,37 +109,23 @@ public String toString()
                 .toString();
     }
 
-    public static <T> JsonResponse<T> execute(TrinoJsonCodec<T> codec, Call.Factory client, Request request, OptionalLong materializedJsonSizeLimit)
+    public static <T> JsonResponse<T> execute(TrinoJsonCodec<T> codec, Call.Factory client, Request request)
     {
         try (Response response = client.newCall(request).execute()) {
             ResponseBody responseBody = requireNonNull(response.body());
             if (isJson(responseBody.contentType())) {
-                String body = null;
                 T value = null;
                 IllegalArgumentException exception = null;
-                try {
-                    if (materializedJsonSizeLimit.isPresent() && (responseBody.contentLength() < 0 || responseBody.contentLength() > materializedJsonSizeLimit.getAsLong())) {
-                        // Parse from input stream, response is either of unknown size or too large to materialize. Raw response body
-                        // will not be available if parsing fails
-                        value = codec.fromJson(responseBody.byteStream());
-                    }
-                    else {
-                        // parse from materialized response body string
-                        body = responseBody.string();
-                        value = codec.fromJson(body);
-                    }
+                MaterializingReader reader = new MaterializingReader(new InputStreamReader(responseBody.byteStream()), 128 * 1024);
+                try (Reader ignored = reader) {
+                    // Parse from input stream, response is either of unknown size or too large to materialize. Raw response body
+                    // will not be available if parsing fails
+                    value = codec.fromJson(reader);
                 }
                 catch (JsonProcessingException e) {
-                    String message;
-                    if (body != null) {
-                        message = format("Unable to create %s from JSON response:\n[%s]", codec.getType(), body);
-                    }
-                    else {
-                        message = format("Unable to create %s from JSON response", codec.getType());
-                    }
-                    exception = new IllegalArgumentException(message, e);
+                    exception = new IllegalArgumentException(format("Unable to create %s from JSON response:\n[%s]", codec.getType(), reader.getInitalString()), e);
```

Comment:
nit: mark in exception message if we have full body or truncated (add `...` if trucated)